### ISSUE TITLE
Integration stop conditions

### ISSUE THREAD

**Issue by nichollsh** [ExPlanetology/aragog#21]
SPIDER has a useful feature where the integration would stop if the surface temperature changed by more than a given amount (in both relative and absolute terms).

This is particularly important for coupling to PROTEUS, because a large change would mean that we should recalculate the atmosphere and therefore the upper flux BC for Aragog.

It would be useful if this condition -- and maybe similar ones on melt fraction, etc -- were implemented into Aragog too.

**Comment by nichollsh** [comment]
This should be possible to do using the `events` keyword of `solve_ivp` to provide it with a function that terminates the integration when these criteria are satisfied.

https://docs.scipy.org/doc/scipy/reference/generated/scipy.integrate.solve_ivp.html#scipy.integrate.solve_ivp

**Comment by djbower** [comment]
Two thoughts on this:

- The general process of stopping and starting time-evolving modules could be managed by PROTEUS directly.  Since PROTEUS knows about the state of all components in the system, it could implement a basic look-ahead algorithm, like linear extrapolation, to predict the duration of the time integration. For Aragog, this could be based on dTsurf/dt.  Then the framework for stopping and starting modules is centralised within PROTEUS, and could be applied across all time-evolving modules (which could eventually include different surface lid/boundary layer models).
- Otherwise, yes, having a solver API that supports events makes life much easier.

**Comment by nichollsh** [comment]
Using a look-ahead algorithm is an interesting idea. It sounds like a simple one would be relatively easy to implement, and might also go some way towards fixing the issues that sometimes crop up with the current time-stepping solution (https://github.com/FormingWorlds/PROTEUS/issues/193).

**Comment by djbower** [comment]
Yes, I think as the model becomes more non-linear, with multiple interacting modules, it probably makes sense for PROTEUS to have a smart module/class that regulates the time-stepping of all sub-components.  This encapsulates the logic and ensure that the coupler itself (PROTEUS) is dealing with this, rather than modules that may be hot-swapped in and out ("separation of concerns").
